controlplane/kubeadm/internal/controllers/status.go

```diff
@@ -423,6 +424,139 @@ func setRemediatingCondition(ctx context.Context, kcp *controlplanev1.KubeadmCon
 	})
 }
 
+func setAvailableCondition(_ context.Context, kcp *controlplanev1.KubeadmControlPlane, etcdIsManaged bool, etcdMembers []*etcd.Member, etcdMembersAgreeOnMemberList bool, etcdMembersAgreeOnClusterID bool, etcdMembersAndMachinesAreMatching bool, machines collections.Machines) {
+	if !kcp.Status.Initialized {
+		v1beta2conditions.Set(kcp, metav1.Condition{
+			Type:    controlplanev1.KubeadmControlPlaneAvailableV1Beta2Condition,
+			Status:  metav1.ConditionFalse,
+			Reason:  controlplanev1.KubeadmControlPlaneNotAvailableV1Beta2Reason,
+			Message: "Control plane not yet initialized",
+		})
+		return
+	}
+
+	if etcdIsManaged && etcdMembers == nil {
+		v1beta2conditions.Set(kcp, metav1.Condition{
+			Type:    controlplanev1.KubeadmControlPlaneAvailableV1Beta2Condition,
+			Status:  metav1.ConditionUnknown,
+			Reason:  controlplanev1.KubeadmControlPlaneAvailableInspectionFailedV1Beta2Reason,
+			Message: "Failed to get etcd members",
+		})
+		return
+	}
+
+	if etcdIsManaged && !etcdMembersAgreeOnMemberList {
+		v1beta2conditions.Set(kcp, metav1.Condition{
+			Type:    controlplanev1.KubeadmControlPlaneAvailableV1Beta2Condition,
+			Status:  metav1.ConditionFalse,
+			Reason:  controlplanev1.KubeadmControlPlaneNotAvailableV1Beta2Reason,
+			Message: "At least one etcd member reports a list of etcd members different than the list reported by other members",
+		})
+		return
+	}
+
+	if etcdIsManaged && !etcdMembersAgreeOnClusterID {
+		v1beta2conditions.Set(kcp, metav1.Condition{
+			Type:    controlplanev1.KubeadmControlPlaneAvailableV1Beta2Condition,
+			Status:  metav1.ConditionFalse,
+			Reason:  controlplanev1.KubeadmControlPlaneNotAvailableV1Beta2Reason,
+			Message: "At least one etcd member reports a cluster ID different than the cluster ID reported by other members",
+		})
+		return
+	}
+
+	if etcdIsManaged && !etcdMembersAndMachinesAreMatching {
+		v1beta2conditions.Set(kcp, metav1.Condition{
+			Type:    controlplanev1.KubeadmControlPlaneAvailableV1Beta2Condition,
+			Status:  metav1.ConditionFalse,
+			Reason:  controlplanev1.KubeadmControlPlaneNotAvailableV1Beta2Reason,
+			Message: "The list of etcd members does not match the list of Machines and Nodes",
+		})
+		return
+	}
+
+	// Determine control plane availability looking at machines conditions, which at this stage are
+	// already surfacing status from etcd member and all control plane pods hosted on every machine.
+	// Note: we intentionally use the number of etcd members for determine the etcd quorum because
+	// etcd members could not match with machines, e.g. while provisioning a new machine.
```

Comment:
```suggestion
	// etcd members might not match with machines, e.g. while provisioning a new machine.
```
Not sure if that's what you mean instead